File: src/main/java/com/example/probation/service/impl/UsersServiceImpl.java
package com.example.probation.service.impl;

import com.example.probation.core.entity.User;
import com.example.probation.event.OnRegistrationCompleteEvent;

import com.example.probation.exception.ForbiddenException;
import com.example.probation.exception.TimeHasExpiredException;
import com.example.probation.exception.NoSuchUserException;
import com.example.probation.exception.TokenNotFoundException;
import com.example.probation.exception.UserNotFoundByTokenException;
import com.example.probation.repository.UsersRepository;
import com.example.probation.service.RoleService;
import com.example.probation.service.TokenService;
import com.example.probation.service.UsersService;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.Calendar;

@Transactional
@Service
@RequiredArgsConstructor
public class UsersServiceImpl implements UsersService {
    public static final Integer ADDITION = 15;
    private final UsersRepository usersRepository;
    private final PasswordEncoder passwordEncoder;
    private final RoleService roleService;
    private final TokenService tokenService;
    private final CustomUserDetailsService detailsService;
    private final ApplicationEventPublisher eventPublisher;

    @Override
    public User registerNewUser(final User user) {
        final var userRole = "USER";
        user.setRoles(Set.of(roleService.getRoleByRoleName(userRole)));
        user.setPassword(passwordEncoder.encode(user.getPassword()));
        eventPublisher.publishEvent(new OnRegistrationCompleteEvent(user));
        return usersRepository.save(user);
    }

    @Override
    public boolean checkEmailExistence(final String email) {
        return usersRepository.findByEmail(email).isPresent();
    }

    @Override
    public boolean checkUsernameExistence(final String username) {
        return usersRepository.findByUsername(username).isPresent();
    }

    @Override
    public void redefineRating(final User winner, final User loser) {
        final List<User> players = new ArrayList<>();
        loser.setRating(calculateLoserRating(loser.getRating()));
        winner.setRating(calculateWinnerRating(winner.getRating()));
        players.add(winner);
        players.add(loser);
        usersRepository.saveAll(players);
    }

    @Override
    public List<User> getUsersForComparison() {
        return usersRepository.getRandomUsers();
    }

    @Override
    public List<User> getTopUsersByRating() {
        return usersRepository.findAllByOrderByRatingDesc();
    }

    @Override
    public Integer calculateWinnerRating(final Integer currentRating) {
        return currentRating + ADDITION;
    }

    @Override
    public Integer calculateLoserRating(final Integer currentRating) {
        return currentRating - ADDITION;
    }

    @Override
    public Optional<User> findByUserName(final String username) {
        return usersRepository.findByUsername(username);
    }

    @Override
    public User getCurrentUser() {
        final String username = detailsService.getCurrentUsername();

        if (username == null) {
            throw new NoSuchUserException("{user.no.such}");
        }

        return findByUserName(username).orElseThrow(ForbiddenException::new);
    }

    @Override
    public Optional<User> getUserByToken(final String token) {
        return Optional.of(tokenService.findByToken(token).orElseThrow(TokenNotFoundException::new).getUser());
    }

    @Override
    public User saveRegisteredUser(final String token) {
        final var verificationToken = tokenService.findByToken(token).orElseThrow(UserNotFoundByTokenException::new);
        final var user = tokenService.getUserByToken(token).orElseThrow();
        final var cal = Calendar.getInstance();

        if ((verificationToken.getExpiryDate().getTime() - cal.getTime().getTime()) <= 0) {
            throw new TimeHasExpiredException();
        }

        user.setId(user.getId());
        user.setEnabled(true);
        usersRepository.save(user);

        return user;
    }
}

File: src/main/kotlin/com/example/probation/service/impl/UsersServiceImpl.kt
package com.example.probation.service.impl

import com.example.probation.core.dto.ChangePasswordDto
import com.example.probation.core.entity.User
import com.example.probation.core.enums.Actions
import com.example.probation.core.enums.Roles
import com.example.probation.event.OnRegistrationCompleteEvent
import com.example.probation.exception.EntityNotFoundException
import com.example.probation.exception.PasswordDoesNotMatchesException
import com.example.probation.exception.TimeHasExpiredException
import com.example.probation.repository.UsersRepository
import com.example.probation.service.KafkaProducerService
import com.example.probation.service.RoleService
import com.example.probation.service.TokenService
import com.example.probation.service.UsersService
import org.springframework.context.ApplicationEventPublisher
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class UsersServiceImpl(
    private val usersRepository: UsersRepository,
    private val passwordEncoder: PasswordEncoder,
    private val roleService: RoleService,
    private val tokenService: TokenService,
    private val detailsService: CustomUserDetailsService,
    private val eventPublisher: ApplicationEventPublisher,
    private val kafkaService: KafkaProducerService
) : UsersService {

    override fun registerNewUser(newUser: User) = newUser.apply {
        password = passwordEncoder.encode(newUser.password)
        roleService.getRoleByRoleName(Roles.USER.role)?.let { role ->
            roles = mutableSetOf(role)
        }
    }.let { usersRepository.save(it) }
        .also {
            eventPublisher.publishEvent(OnRegistrationCompleteEvent(it))
            val currentUsername = detailsService.getCurrentUsername()
            kafkaService.send(Actions.REGISTER.action, currentUsername, it.username)
        }

    override fun redefineRating(winner: User, loser: User) {
        val players: MutableList<User> = ArrayList()
        loser.rating = calculateLoserRating(loser.rating)
        winner.rating = calculateWinnerRating(winner.rating)
        players.add(winner)
        players.add(loser)
        usersRepository.saveAll(players)
        kafkaService.send(Actions.CHOOSE.action, detailsService.getCurrentUsername(), winner.username)
    }

    override fun getUsersForComparison() = usersRepository.getRandomUsers()

    override fun getTopUsersByRating() = usersRepository.findAllByOrderByRatingDesc()

    override fun calculateWinnerRating(currentRating: Int) = currentRating + ADDITION

    override fun calculateLoserRating(currentRating: Int) = currentRating - ADDITION

    override fun findByUserName(username: String) =
        usersRepository.findByUsername(username) ?: throw EntityNotFoundException("{entity.notfound}")

    override fun getUserByToken(token: String) =
        tokenService.findByToken(token)?.user

    override fun saveRegisteredUser(token: String) =
        tokenService.findByToken(token)?.let { verificationToken ->
            if (verificationToken.expiryDate?.isAfter(LocalDateTime.now()) != true) {
                throw TimeHasExpiredException("{time.expired}")
            }
            tokenService.getUserByToken(token)?.enable(true)?.let {
                usersRepository.save(it)
            }.also {
                it?.username?.let { name -> kafkaService.send(Actions.CONFIRM.action, name) }
            }
        }

    private fun User.enable(isEnabled: Boolean) = apply { enabled = isEnabled }

    override fun getCurrentUser() =
        findByUserName(detailsService.getCurrentUsername())

    override fun checkEmailExistence(email: String) =
        usersRepository.findByEmail(email) != null

    override fun checkUsernameExistence(username: String) =
        usersRepository.findByUsername(username) != null

    override fun getUserById(id: Long): User =
        usersRepository.getUserById(id) ?: throw EntityNotFoundException("{entity.not.found}")


    override fun blockUser(userId: Long) {
        getUserById(userId).enable(false).let {
            usersRepository.save(it)
        }.also {
            kafkaService.send(Actions.BLOCK.action, detailsService.getCurrentUsername(), it.username)
        }
    }

    override fun unblockUser(userId: Long) {
        getUserById(userId).enable(true).let {
            usersRepository.save(it)
        }.also {
            kafkaService.send(Actions.UNBLOCK.action, detailsService.getCurrentUsername(), it.username)
        }
    }

    override fun deleteUser(userId: Long) {
        getUserById(userId).let {
            tokenService.deleteTokenByUser(it)
            usersRepository.deleteById(userId)
            val currentUsername = detailsService.getCurrentUsername()
            kafkaService.send(Actions.DELETE.action, currentUsername, it.username)
            it.username?.let { name ->
                kafkaService.send(Actions.DELETED.action, name, currentUsername)
            }
        }
    }

    override fun checkUniqueNewName(newName: String, id: Long) =
        getUserById(id).run {
            newName == username || !checkUsernameExistence(newName)
        }

    override fun checkUniqueNewEmail(newEmail: String, id: Long) =
        getUserById(id).run {
            newEmail == email || !checkEmailExistence(newEmail)
        }

    override fun changePersonalData(user: User) {
        user.id?.let { userId ->
            getUserById(userId).apply {
                id = user.id
                username = user.username
                description = user.description
                email = user.email
            }.let {
                usersRepository.save(it)
            }.also {
                it.username?.let { name -> kafkaService.send(Actions.EDIT.action, name) }
            }
        }
    }

    override fun changePassword(changePasswordDto: ChangePasswordDto) {
        getUserById(changePasswordDto.id).apply {
            if (passwordEncoder.matches(changePasswordDto.oldPassword, password)) {
                password = passwordEncoder.encode(changePasswordDto.newPassword)
                usersRepository.save(this)
            } else {
                throw PasswordDoesNotMatchesException("{password.not.matches}")
            }
        }.also {
            it.username?.let { name ->
                kafkaService.send(Actions.CHANGE_PASSWORD.action, name)
            }
        }
    }

    companion object {
        const val ADDITION = 15
    }
}
